File: 백준/Silver/15650. N과 M （2）/N과 M （2）.cc
#include <iostream>
#include <vector>

using namespace std;

template <typename S> void println(const S& s) {

	for (const auto& p : s) {
		cout << p << "\n";
	}
}

int N, M;
vector<string> vec;
vector<bool> used;

void backtrack(int len, string ans, int idx) {
	if (len == M) {
		string str = ans.substr(1);
		vec.push_back(str);
		return;
	}

	for (int i = idx + 1; i <= N; i++) {
		if (!used[i]) {
			used[i] = true;
			backtrack(len + 1, ans + " " + to_string(i), i);
			used[i] = false;
		}
	}
}

int main() {

	scanf("%d %d", &N, &M);

	used.resize(N + 1, false);

	backtrack(0, "", 0);

	println(vec);

	return 0;
}

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

/**
 * Main
 */
public class Main {

    private static int N, M;
    private static boolean[] used;
    private static StringBuilder sb;

    public static void main(String[] args) throws IOException{
        //System.setIn(new FileInputStream("input.txt"));

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        used = new boolean[N + 1];
        sb = new StringBuilder();
        
        backtrack(0, "", 0);

        System.out.println(sb);
    }

    private static void backtrack(int len, String ans, int idx){
        if(len == M){
            sb.append(ans.trim()).append("\n");
            return;
        }
        for (int i = idx+1; i <= N; i++) {
            used[i] = true;
            backtrack(len + 1, ans + " " + i, i);
            used[i] = false;
        }
    }   
}